dll/win32/kernel32/client/appcache.c

```diff
@@ -309,23 +309,78 @@ BasepCheckBadapp(IN HANDLE FileHandle,
 }
 
 /*
- * @unimplemented
+ * @implemented
  */
-VOID
+BOOL
 WINAPI
 BaseDumpAppcompatCache(VOID)
 {
-    STUB;
+    NTSTATUS Status;
+
+    Status = NtApphelpCacheControl(ApphelpCacheServiceDump, NULL);
+    return NT_SUCCESS(Status);
 }
 
+BOOLEAN
+WINAPI
+IsCallerAdminOrSystem(VOID)
+{
+    SID_IDENTIFIER_AUTHORITY SidAuthority = {SECURITY_NT_AUTHORITY};
+    PSID pSid = NULL;
+    NTSTATUS Status;
+    HMODULE hAdvapi32;
+    BOOL (WINAPI* pCheckTokenMembership)(HANDLE TokenHandle, PSID SidToCheck, PBOOL IsMember);
+    BOOL Allowed = FALSE;
+
+    hAdvapi32 = LoadLibraryW(L"advapi32.dll");
+    if (!hAdvapi32)
+        return FALSE;
+
+    pCheckTokenMembership = (PVOID)GetProcAddress(hAdvapi32, "CheckTokenMembership");
+
+    if (pCheckTokenMembership)
+    {
+        Status = RtlAllocateAndInitializeSid(&SidAuthority, 2, SECURITY_BUILTIN_DOMAIN_RID,
+                                             DOMAIN_ALIAS_RID_ADMINS, 0, 0, 0, 0, 0, 0, &pSid);
+        if (NT_SUCCESS(Status))
+        {
+            pCheckTokenMembership(NULL, pSid, &Allowed);
+
+            RtlFreeSid(pSid);
+        }
+
+        if (!Allowed)
+        {
+            Status = RtlAllocateAndInitializeSid(&SidAuthority, 1, SECURITY_LOCAL_SYSTEM_RID,
+                                                 0, 0, 0, 0, 0, 0, 0, &pSid);
+            if (NT_SUCCESS(Status))
+            {
+                pCheckTokenMembership(NULL, pSid, &Allowed);
+
+                RtlFreeSid(pSid);
+            }
+        }
+    }
+
+    FreeLibrary(hAdvapi32);
+
+    return Allowed;
+}
+
+
 /*
- * @unimplemented
+ * @implemented
  */
-VOID
+BOOL
 WINAPI
 BaseFlushAppcompatCache(VOID)
 {
-    STUB;
+    NTSTATUS Status;
+    if (!IsCallerAdminOrSystem())
+        return FALSE;
+
+    Status = NtApphelpCacheControl(ApphelpCacheServiceFlush, NULL);
```

Comment:
To extend my previous comment about insecure advapi32 loading, the whole code path here looks insecure to me. It's not up to the caller to check it has proper rights, but up to the kernel to check its caller is granted enough rights.
Looking at our NtApphelpCacheControl() implementation, it doesn't seem to check any rights?
(according to CVE-2015-0002, it's indeed up to the kernel to perform the check!)

Reply:
Ok so I'll move this entire check to NtApphelpCacheControl then. (And drop the usermode path).
I will save this for a later PR.

Reply:
I added a comment in NtApphelpCacheControl to fix this later.